Fix: Improve text encoding for shell output in VSCode preview (#6178)

LIHUA919 opened the pull request:
## 🐛 Problem

Users running commands with non-ASCII characters (like Russian text "пример") in Windows/WSL environments experience garbled text in VSCode's shell preview window, with Unicode replacement characters (�) appearing instead of the actual text.

**Issue**: https://github.com/openai/codex/issues/6178

## 🔧 Root Cause

The issue was in `StreamOutput<Vec<u8>>::from_utf8_lossy()` method in `codex-rs/core/src/exec.rs`, which used `String::from_utf8_lossy()` to convert shell output bytes to strings. This function immediately replaces any invalid UTF-8 byte sequences with replacement characters, without attempting to decode using other common encodings.

In Windows/WSL environments, shell output often uses encodings like:

- Windows-1252 (common Windows encoding)
- Latin-1/ISO-8859-1 (extended ASCII)

## 🛠️ Solution

Replaced the simple `String::from_utf8_lossy()` call with intelligent encoding detection via a new `bytes_to_string_smart()` function that tries multiple encoding strategies:

1. **UTF-8** (fast path for valid UTF-8)
2. **Windows-1252** (handles Windows-specific characters in 0x80-0x9F range)
3. **Latin-1** (fallback for extended ASCII)
4. **Lossy UTF-8** (final fallback, same as before)

## 📁 Changes

### New Files

- `codex-rs/core/src/text_encoding.rs` - Smart encoding detection module
- `codex-rs/core/tests/suite/text_encoding_fix.rs` - Integration tests

### Modified Files

- `codex-rs/core/src/lib.rs` - Added text_encoding module
- `codex-rs/core/src/exec.rs` - Updated StreamOutput::from_utf8_lossy()
- `codex-rs/core/tests/suite/mod.rs` - Registered new test module

## ✅ Testing

- **5 unit tests** covering UTF-8, Windows-1252, Latin-1, and fallback scenarios
- **2 integration tests** simulating the exact Issue #6178 scenario
- **Demonstrates improvement** over the previous `String::from_utf8_lossy()` approach

All tests pass:

```bash
cargo test -p codex-core text_encoding
cargo test -p codex-core test_shell_output_encoding_issue_6178
```

## 🎯 Impact

- ✅ **Eliminates garbled text** in VSCode shell preview for non-ASCII content
- ✅ **Supports Windows/WSL environments** with proper encoding detection
- ✅ **Zero performance impact** for UTF-8 text (fast path)
- ✅ **Backward compatible** - UTF-8 content works exactly as before
- ✅ **Handles edge cases** with robust fallback mechanism

## 🧪 Test Scenarios

The fix has been tested with:

- Russian text ("пример")
- Windows-1252 quotation marks (""test")
- Latin-1 accented characters ("café")
- Mixed encoding content
- Invalid byte sequences (graceful fallback)

## 📋 Checklist

- [X] Addresses the reported issue
- [X] Includes comprehensive tests
- [X] Maintains backward compatibility
- [X] Follows project coding conventions
- [X] No breaking changes

LIHUA919 committed:
fix: improve text encoding for shell output in VSCode preview (#6178)

   Replace String::from_utf8_lossy() with intelligent encoding detection
   in StreamOutput::from_utf8_lossy() to fix garbled text in VSCode shell
   preview when running commands with non-ASCII characters in Windows/WSL.

   The new bytes_to_string_smart() function tries multiple encoding strategies:
   1. UTF-8 (fast path for valid UTF-8)
   2. Windows-1252 (common in Windows environments)
   3. Latin-1 (fallback for extended ASCII)
   4. Lossy UTF-8 (final fallback)

LIHUA919 committed:
Fix code formatting for CI

   Apply cargo fmt to fix formatting issues:
   - Reorder module declarations alphabetically
   - Fix line length and spacing in text_encoding.rs
   - Add newline at end of file

LIHUA919 committed:
Permit ANSI escape codes in encoding detection

  Allow ESC (\x1b) character in Windows-1252 and Latin-1 detection
  to support colored shell output with ANSI escape sequences.

  This addresses the code review feedback that typical shell output
  embeds ANSI colour sequences alongside Windows-1252 characters,
  and the previous strict control character check was causing these
  to fall back to lossy UTF-8 conversion.

joshka-oai commented on codex-rs/core/src/text_encoding.rs:
```diff
@@ -0,0 +1,151 @@
+
+    // Validate that the result makes sense (contains reasonable characters)
+    if result.chars().any(|c| c.is_control() && c != '\n' && c != '\r' && c != '\t') {
+        return None;
```
This is still better than what was there before. The problem mentioned by codex is correct. This PR narrows the current failure mode to 1252/latin + ANSI sequences. Ideally shell commands that are run non-interactively should omit formatting when possible, so it's likely this is a much smaller portion of the whole.

That said, a fairly trivial fix for this might be to add \x1b to the allowed character lists here.

joshka-oai reviewed (commented):
General approval with the approach and code changes from a visual inspection. I haven't yet tested this however, but will later this week.

joshka-oai committed:
use encoding_rs and chardetng

We already have encoding_rs in our dependency list, and it's a widely
used package (275M downloads). Chardetng is by the same developer.

joshka-oai committed:
Merge branch 'main' into fix/text-encoding-6178

joshka-oai committed:
ignore spelling on test cases

joshka-oai committed:
Fix problem with smart punctuation overlaps

joshka-oai committed:
uncap cp866 problem. reformat comments

joshka-oai commented on codex-rs/core/src/text_encoding.rs:
```diff
@@ -0,0 +1,199 @@
+    0x95, // • (bullet)
+    0x96, // – (en dash)
+    0x97, // — (em dash)
+    0x99, // ™ (trade mark sign)
```
I'm going to push back against this one for now - we could chase all the various characters here. Let's wait for an actual problem to occur to re-review which characters should be in this list.

nos1609 commented:
@joshka-oai The only doubt actually from my point of view is the lack of Windows 1251 codepage. The more kind-of "universal" would be to handle the Linux's [LatArCyrHeb](https://developer.ibm.com/articles/l-u-cyr/) approch to cover non Latin-charaters output, especially cyrillic ones.

The Windows 1252 looks like not covering the ArCyrHeb segments of the caracter maps.

Just a small "hello world" example
The word "Привет" in Windows 1252 would be rendered as "ÐŸÑ€Ð¸Ð²ÐµÑ‚" which is not really human-legible as "Hello" because of lack of cyrillic characters from Windows 1251.

https://learn.microsoft.com/windows/win32/intl/code-page-identifiers

joshka-oai committed:
Add a couple of extra tests for 1251 and 1252 text

joshka-oai committed:
Add a boatload of tests
